File: asr/harmony/sdk-dingqiao/src/main/ets/com/amphion/dingqiao/diarization/SpeakerDiarizationSpeakerIndex.ts
export const UNASSIGNED_SPEAKER_INDEX: number = -1;

export function speakerIndexFromInternalId(speakerId: string, _maxSpeakers?: number): number {
  if (!speakerId.startsWith('S')) return UNASSIGNED_SPEAKER_INDEX;
  const value = Number(speakerId.substring(1));
  return Number.isInteger(value) && value > 0 ?
    value - 1 : UNASSIGNED_SPEAKER_INDEX;
}

export function speakerIndexesFromInternalIds(speakerIds: string[], maxSpeakers?: number,
  preserveUnassigned: boolean = false): number[] {
  const indexes: number[] = [];
  for (let index = 0; index < speakerIds.length; index++) {
    const value = speakerIndexFromInternalId(speakerIds[index], maxSpeakers);
    if (value >= 0 && indexes.indexOf(value) < 0) {
      indexes.push(value);
    } else if (value < 0 && preserveUnassigned && indexes.indexOf(UNASSIGNED_SPEAKER_INDEX) < 0) {
      indexes.push(UNASSIGNED_SPEAKER_INDEX);
    }
  }
  return indexes;
}

File: asr/android/sdk-dingqiao/src/main/java/com/amphion/dingqiao/diarization/SpeakerDiarizationSpeakerIndex.kt
package com.amphion.dingqiao.diarization

internal fun speakerIndexFromInternalId(speakerId: String, maxSpeakers: Int): Int {
    if (!speakerId.startsWith("S")) return -1
    val oneBased = speakerId.substring(1).toIntOrNull() ?: return -1
    return if (oneBased in 1..maxSpeakers) oneBased - 1 else -1
}

internal fun speakerIndexesFromInternalIds(
    speakerIds: List<String>,
    maxSpeakers: Int,
    excludeUnknown: Boolean,
): List<Int> = speakerIds.map { speakerIndexFromInternalId(it, maxSpeakers) }
    .filter { !excludeUnknown || it >= 0 }
    .distinct()
